pkg/controller/shared_state.go

```diff
@@ -1345,13 +1373,152 @@ func (c *SharedState) createReplicas(ctx context.Context, remainingReplicas int,
+		}
+
+		if events.isEmpty() {
+			lastTime = time.Now()
+			delay = newEvent.timestamp.Sub(lastTime) - expLatency
+			alarm = time.NewTimer(delay)
+		}
+
+		events.add(newEvent)
+
+		eventMap[newFailureInfo.volumeName].tryRemove()
+		eventMap[newFailureInfo.volumeName] = newEvent
+
+	NoNewFailureLoop:
+		for {
+			select {
+			case newFailureInfo = <-c.eventsToPersistQueue:
+				break NoNewFailureLoop
+			case newVolumeName := <-c.eventsToUnpersistQueue:
+				oldEvent := eventMap[newVolumeName]
+				if oldEvent == nil {
+					continue NoNewFailureLoop
+				}
+				delete(eventMap, newVolumeName)
+				newFailureInfo = ReplicaAttachmentFailureInfo{}
+				if oldEvent.next == oldEvent {
+					// no more events to refresh!
+					events.clear()
+					klog.V(2).Infof("Deactivating the event refresher")
+
+					// allow resources associated with the alarm to be released
+					if !alarm.Stop() {
+						<-alarm.C
+					}
+					alarm = nil
+
+					// prevent additional event un-persist requests
+					unmake(&c.eventsToUnpersistQueue)
+
+					// wait until a new event needs refreshing
+					newFailureInfo = <-c.eventsToPersistQueue // during this period of time, `c.unpersistAttachmentFailure` becomes a no-op nil-check
+
+					// allow new event un-persist requests
+					c.eventsToUnpersistQueue = make(chan string, maxUnpersistEventConcurrency)
+					klog.V(2).Infof("Reactivating the event refresher")
+
+					break NoNewFailureLoop
+				} else {
+					oldEvent.remove()
+				}
+			case <-alarm.C:
+				currTime := time.Now()
+				currLatency := expLatency + currTime.Sub(lastTime) - delay
+				expLatency = (expLatency*3 + currLatency) / 4 // update the expected latency with a 25% expontially-weighted moving average
+				lastTime = currTime
+				for {
+					for _, object := range events.curr.objects { // can we keep this as is? or do we need to check / keep track of whether each pod is still relevant
+						c.eventRecorder.Event(object, v1.EventTypeWarning, consts.ReplicaAttachmentFailedEvent, events.curr.message)
+					}
+					events.curr.timestamp = events.curr.timestamp.Add(eventTTL)
+					events.next()
+					delay = time.Until(events.curr.timestamp) - expLatency
+					if delay >= eventOverlapVariance {
+						break
+					}
+					// events which are really close together will be processed at the same, preventing the overuse of sleep()
+					// event refresher's responsiviness may be more variable, so we'll increase its channel buffers as needed
+				}
+				alarm.Reset(delay)
+			}
+		}
+	}
+}
+
+func unmake[T any](channelPtr *chan T) {
+	channel := *channelPtr
+	*channelPtr = nil
+	for {
+		select {
+		case <-channel:
+			break
+		default:
+			close(channel)
+			return
+		}
+	}
+}
+
+func trySend[T any](channel chan T, value T) {
+	if channel != nil {
+		trySendNonNil(channel, value)
+	}
+}
+
+func trySendNonNil[T any](nonNilChannel chan T, value T) {
+	defer func() { _ = recover() }()
```

Comment:
We shouldn't rely on ignoring panics to handle the race between `unmake` and `trySend`.

Reply:
please elaborate on the potential pitfalls of doing this for this scenario, as sending to a non-nil channel only panics if the channel is closed; the fact that go panics when attempting to send to a closed channel is worked around in this way in [docker & k8s](https://github.com/search?q=%28org%3Akubernetes+OR+org%3Adocker+OR+org%3Amoby++%29+%2F%5Csrecover%5C%28%5B%5Cs%5CS%5Cr%5Cn%5D*%5C%7D%5C%28%5C%29%5Cn*%5Cs*%5Cw%2B+%3C-+%5Cw%2B%2F+-path%3A**test**&type=code)

Reply:
Panics should be reserved for genuinely unexpected scenarios and code should be structured to avoid them. If we expect them by design, that can mask cases were the channel was closed unexpectedly leading to undefined behavior.

Reply:
Sending to a closed channel is not something that is _expected_ to happen, but there _are_ race conditions in which the channel might be unexpectedly closed (really, really small chance).
The moby / docker project mitigates this in the same way: https://github.com/moby/moby/issues/37735
`trySend` is also only used for un-persisting attachment failures, and if such a channel was unintentionally closed due to a programming mistake, the consequences would be very insignificant...
Perhaps we could move `trySend` and `trySendNonNil` inside the un-persist function to restrict its use?
And/or log a warning when an error is encountered?

Reply:
Why close the channel at all? It doesn't take a significant amount of resources and the process itself is long-lived (i.e. it will only stop if it panics or the pod is explicitly terminated e.g. for node shutdown).

As I mentioned in [another comment](https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1781#discussion_r1258925137), I believe we can simplify this loop by eliminating the "inactive" state. This would essentially require having the channel open always. Go channels are designed as the primary concurrency mechanism support multiple producer/single (or multiple) consumer scenarios. Though not "lock-free", using channels is designed to be fast and scalable, so optimizing to avoid sending on a channel isn't really necessary.